packages/tailwindcss/src/index.ts

```diff
@@ -143,40 +175,98 @@ export function compile(css: string): {
   })
 
   // Replace `@apply` rules with the actual utility classes.
-  if (css.includes('@apply')) {
-    walk(ast, (node, { replaceWith }) => {
-      if (node.kind === 'rule' && node.selector[0] === '@' && node.selector.startsWith('@apply')) {
-        let candidates = node.selector
-          .slice(7 /* Ignore `@apply ` when parsing the selector */)
-          .trim()
-          .split(/\s+/g)
-
-        // Replace the `@apply` rule with the actual utility classes
-        {
-          // Parse the candidates to an AST that we can replace the `@apply` rule with.
-          let candidateAst = compileCandidates(candidates, designSystem, {
-            onInvalidCandidate: (candidate) => {
-              throw new Error(`Cannot apply unknown utility class: ${candidate}`)
-            },
-          }).astNodes
-
-          // Collect the nodes to insert in place of the `@apply` rule. When a
-          // rule was used, we want to insert its children instead of the rule
-          // because we don't want the wrapping selector.
-          let newNodes: AstNode[] = []
-          for (let candidateNode of candidateAst) {
-            if (candidateNode.kind === 'rule' && candidateNode.selector[0] !== '@') {
-              for (let child of candidateNode.nodes) {
-                newNodes.push(child)
+  if (containsAtApply) {
+    walk(ast, (root) => {
+      if (root.kind !== 'rule') return WalkAction.Continue
+
+      // It's possible to `@apply` user-defined classes. We need to make sure
+      // that we never run into a situation where we are eventually applying
+      // the same class that we are currently processing otherwise we will end
+      // up in an infinite loop (circular dependency).
+      //
+      // This means that we need to track the current node as a candidate and
+      // error when we encounter it again.
+      let rootAsCandidate = root.selector.slice(1)
+
+      walk(root.nodes, (node, { replaceWith }) => {
+        if (
+          node.kind === 'rule' &&
+          node.selector[0] === '@' &&
+          node.selector.startsWith('@apply')
+        ) {
+          let candidates = node.selector
+            .slice(7 /* Ignore `@apply ` when parsing the selector */)
```

Comment:
I really enjoy the comments in your code, you could have use
```js
.slice('@apply '.length)
```
instead of
```js
.slice(7 /* Ignore `@apply ` when parsing the selector */)
```
Off course the latest is slightly faster & shorter (if you omit the comment).

In the end, I prefer your version.
Longer with the comment but once you read it once, it feels right 🤓.

Reply:
Yep we focus on performance and re-allocating and re-computing the length is unnecessary work.